middle_end/flambda/un_anf.ml

```diff
@@ -114,104 +152,98 @@ let make_var_info (clam : Clambda.ulambda) : var_info =
       ignore_int_array us_index_blocks;
-      Array.iter loop us_actions_blocks;
+      Array.iter (loop ~depth) us_actions_blocks;
       ignore_debuginfo dbg
     | Ustringswitch (cond, branches, default) ->
-      loop cond;
+      loop ~depth cond;
       List.iter (fun (str, branch) ->
           ignore_string str;
-          loop branch)
+          loop ~depth branch)
         branches;
-      Option.iter loop default
+      Option.iter (loop ~depth) default
     | Ustaticfail (static_exn, args) ->
       ignore_int static_exn;
-      List.iter loop args
+      List.iter (loop ~depth) args
     | Ucatch (static_exn, vars, body, handler) ->
       ignore_int static_exn;
       ignore_params_with_value_kind vars;
-      loop body;
-      loop handler
+      loop ~depth body;
+      loop ~depth handler
     | Utrywith (body, var, handler) ->
-      loop body;
+      loop ~depth body;
       ignore_var_with_provenance var;
-      loop handler
+      loop ~depth handler
     | Uifthenelse (cond, ifso, ifnot) ->
-      loop cond;
-      loop ifso;
-      loop ifnot
+      loop ~depth cond;
+      loop ~depth ifso;
+      loop ~depth ifnot
     | Usequence (e1, e2) ->
-      loop e1;
-      loop e2
+      loop ~depth e1;
+      loop ~depth e2
     | Uwhile (cond, body) ->
-      loop cond;
-      loop body
+      loop ~depth:(depth + 1) cond;
+      loop ~depth:(depth + 1) body
     | Ufor (var, low, high, direction_flag, body) ->
       ignore_var_with_provenance var;
-      loop low;
-      loop high;
+      loop ~depth low;
+      loop ~depth high;
       ignore_direction_flag direction_flag;
-      loop body
+      loop ~depth:(depth + 1) body
     | Uassign (var, expr) ->
-      assigned_vars := V.Set.add var !assigned_vars;
-      loop expr
+      add_assignment t var;
+      loop ~depth expr
     | Usend (meth_kind, e1, e2, args, dbg) ->
       ignore_meth_kind meth_kind;
-      loop e1;
-      loop e2;
-      List.iter loop args;
+      loop ~depth e1;
+      loop ~depth e2;
+      List.iter (loop ~depth) args;
       ignore_debuginfo dbg
     | Uunreachable ->
       ()
   in
-  loop clam;
-  let linear =
-    V.Tbl.fold (fun var n acc ->
-        assert (n >= 1);
-        if n = 1 && not (V.Set.mem var !assigned_vars)
-        then V.Set.add var acc
-        else acc)
-      t V.Set.empty
-  in
-  let assigned = !assigned_vars in
-  let used =
-    (* This doesn't work transitively and thus is somewhat restricted.  In
-       particular, it does not allow us to get rid of useless chains of [let]s.
-       However it should be sufficient to remove the majority of unnecessary
-       [let] bindings that might hinder [Cmmgen]. *)
-    V.Tbl.fold (fun var _n acc -> V.Set.add var acc)
-      t assigned
+  loop ~depth:0 clam;
+  let linear_let_bound_vars, used_let_bound_vars, assigned =
+    V.Tbl.fold (fun var desc ((linear, used, assigned) as acc) ->
+      match desc.uses with
+      | Zero -> acc
```

Comment:
If I understand correctly, the previous code asserted that this case never happened. We might want to preserve either the assertion or a comment saying that this is unexpected.

Reply:
The previous code did not add variables to the table until it saw a use, so the zero case was impossible. The new code adds variables to the table when it sees their definition so the zero case happens for any unused variable.

Reply:
Ah, of course, I missed that.